### ISSUE TITLE
Fix Recurring Events

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-api#1583]
---
### Note

This is a very important feature to implement. 
1. We need someone with either demonstrable:
    1. experience with GraphQL
    2. activity in their GitHub account with a focus on TypeScript
1. This is not an issue for a first time attempt at the technology. 
    1. The first PR needs to be created within a week
    1. All PRs must be completed by the end of February
1. Read through the comments in the closed PR about the progress and why the approach was selected. There are further details below.

In summary:
1. This is not for a newbie
1. This is not for someone with limited time over the next month
---

**Describe the bug**
The creation of a recurring event only creates a single event in the API clients.

**To Reproduce**

1. Create a recurring event
3. Only a single instance is visible

**Expected behavior**

1. You'll need to update or create new mutations that include start and end date for the event recurrence, not just the event itself
1. There will need to be a query created or updated that would return all the calendar entries for a particular time frame including the recurring ones.
1. We will need the ability to edit any recurrence of the event at any time that the event occurs. The edits would need to be valid for:
    1. All instances in the series
    2. Just the instance in question
    4. The instance in question and all those in future
1. Administrators will need to be able to track attendance for any type of event. Therefore you will need to think of ways to create permanence for historical recurring events for this purpose
1. We don't want the creation of an infinite number of event calendar entries if a recurring event has no end date
1. Other functionality must not be affected
1. You must write tests to validate your work

**Actual behavior**
- See above

**Screenshots**
- N/A

**Additional details**
1. Work with @xoldyckk on evaluating your solution here and in the subsequent PR
1. This is a prerequisite for our menu of project ideas for GSoC 2024
1.  When this issue is resolved we'll need to create another one to support custom recurrences, such as `3rd week of the month`, similar to what is available in Google Calendar
1. We almost completed the work for Weekly recurring when the assignee @Community-Programmer was unable to continue due to exam time constraints. 
    1. The first PR for this issue will need to build on the work in this abandoned PR as it was almost approved for merging. 
        1. https://github.com/PalisadoesFoundation/talawa-api/pull/1738
    1. You will need to submit a plan as described here
         1. https://github.com/PalisadoesFoundation/talawa-api/pull/1738#issuecomment-1910724294
1. You will need to split this issue into multiple PRs, each with custom options provided by the RRULE package
    1. Weekly recurring
    2. Daily recurring
    3. Monthly recurring
    5. Annually recurring

**Potential internship candidates**
- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by Community-Programmer** [comment]
Please assign me , i will work on this issue

**Comment by palisadoes** [comment]
Please read the requirements. They have been frequently updated.

**Comment by Community-Programmer** [comment]
Thank you for providing clarification on the requirements for this issue. . I have carefully reviewed the outlined steps and understand the expectations for the solution.

But, I'd like to request to give me some time to thoroughly investigate the issue. To ensure a comprehensive resolution, I plan to go through the codebase, understand the program's flow, and conduct a deep analysis of the problem before diving into the implementation.

Rest assured, I'll keep you posted on my progress, and if there are any points where I require clarification or assistance, I won't hesitate to reach out for your guidance.

Thank You

**Comment by palisadoes** [comment]
Please work with @xoldyckk on this for updates, advice and general comments

**Comment by palisadoes** [comment]
We don't want the creation of an infinite number of event calendar entries if a recurring event has no end date

**Comment by Community-Programmer** [comment]
> We don't want the creation of an infinite number of event calendar entries if a recurring event has no end date

@palisadoes 

Yeah, got it. Will handle this with any of the following solutions:

- Giving user confirmation or warning
- Default maximum recurrence or system-defined maximum

I'm currently going through the issue , and I'm actively exploring all possible solutions. Once I've completed my analysis, I'll reach out @xoldyckk  to discuss the process I plan to follow

**Comment by palisadoes** [comment]
1. You may want to consider a separate collection for recurring events with just a start and end date. Then let the web app populate the calendar based on that information. That would eliminate the infinite problem.
1. The challenge would then be editing any event in the series (only this event, all events, this and future events), which could be solved by a linked list of recurring events each with is own start and stop date.

Whether graphql could solve this is another matter.

**Comment by Community-Programmer** [comment]
> 1. You may want to consider a separate collection for recurring events with just a start and end date. Then let the web app populate the calendar based on that information. That would eliminate the infinite problem.
> 2. The challenge would then be editing any event in the series (only this event, all events, this and future events), which could be solved by a linked list of recurring events each with is own start and stop date.
> 
> Whether graphql could solve this is another matter.

@palisadoes @xoldyckk 

Instead of implementing a linked list of recurring events, I have another, better, and more effective solution.

1. We will create a separate collection for recurring events that will be linked to the Event schema. 
2. When an event is created, we will check whether it is recurring or not. If it is, we will create instances for recurring events. 
3. This approach will ensure smoother and more reliable event editing for present and future events. 
4. Additionally, we will get a unique ID to each event, which will be helpful for efficient editing of events.


![image](https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/d57b1e80-f4f7-4406-8394-8837f95c0eda)


### Here is a rough idea how we can implement this:

**Mongoose Schema**

```javascript
const mongoose = require('mongoose');
const { Schema } = mongoose;

const recurringEventInstanceSchema = new Schema({
  eventId: { type: Schema.Types.ObjectId, ref: 'Event', required: true },
  start: { type: Date, required: true },
  end: { type: Date, required: true },
});

const eventSchema = new Schema({
  title: { type: String, required: true },
  start: { type: Date, required: true },
  end: { type: Date, required: true },
  frequency: { type: String, required: true },
  interval: { type: Number, required: true },
  recurring: {type:  boolean}, //Either true or false
});

const RecurringEventInstanceModel = mongoose.model('RecurringEventInstance', recurringEventInstanceSchema);
const EventModel = mongoose.model('Event', eventSchema);
``` 
**GraphQL mutation**

```javascript
Mutation: {
    createEvent: async (_, { event }) => {
      try {
        const createdEvent = await EventModel.create(event);
        
        // If the event is recurring, generate and store individual instances
        if (createdEvent.recurring) {
          await generateRecurringEventInstances(createdEvent);
        }
        
        return createdEvent;
      } catch (error) {
        throw new Error('Error creating event');
      }
    }
};
``` 
According to this approach, we can generate a query for updating individual events with the unique event ID. I believe this will be the most efficient solution to solve this issue. 

What is your opinion on this?

**Comment by palisadoes** [comment]
1. FYI - We already have an `Event` collection with some of the fields mentioned
    1. https://docs.talawa.io/docs/schema/objects/event
1. This post describes what Google Calendar does. 
    1. They have the equivalent of `recurringEventInstanceSchema` 
    2. They create the entire series of events upon creation of the recurring event
    1. One of the comments states that Google's recurring events all end in the year 2100. 
    3. They have a mechanism for creating new parents in `recurringEventInstanceSchema` if any part of a recurring instance is updated. eg:
        1. Only this event
        2. All events
        4. This and future events
    5. We should use this approach.
    6. So in your case all the frequency tracking fields would be in the `recurringEventInstanceSchema` it would seem.
    7. The existing `Event` collection would have a pointer back to the `recurringEventInstanceSchema`
    8. https://stackoverflow.com/questions/50204299/recurring-events-database-model
1. You will need to accommodate the customizations that Google has for recurring events
    1. Weekly on any combination of days
    2. Monthly on a specific day(s), and week(s) of the month
        1. For the sake of simplicity, we should do this as a follow up issue as soon as this one closes. I don't want to have tricky timing delay the overall goal.

Your mutation logic for handling edits could be tricky, including. 
1. Adjusting the day that the event falls on from date X onward.
    1. We should also leave this for a follow up issue for simplicity. Open it up as soon as the PR for the basic functionality is complete.

**Comment by Community-Programmer** [comment]
@xoldyckk @palisadoes 

Thanks for providing detailed information. I've reviewed the Google Calendar approach and the `recurringEventInstanceSchema`, and I agree with your proposal to follow a similar model for our system.

I'll prioritize implementing the basic functionality first, aligning with Google Calendar's method. Once that's completed and the initial PR is merged, we can address the additional complexities and customizations, such as accommodating different recurrence patterns.

I'll keep you posted on the progress, and we can open follow-up issues as needed.

Thanks for your guidance, and let's move forward with this plan

**Comment by Community-Programmer** [comment]
@xoldyckk  @palisadoes 

I am almost prepared to create a pull request for this issue but require some clarifications.

I've included the same fields for recurring events as those for regular events, adding only one more field: 'parentEvent.' Everything else, including both GraphQL input and type, remains unchanged. Do I need to make any additional additions or removals?

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/eec25398-3db4-4c3b-a9d4-2aeb5db1beb9)

**recurring Event Model**

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/f1b58bde-9e2a-4089-8e07-bb1b3533e435)

**Comment by palisadoes** [comment]
@Community-Programmer 

1. What is the logic of operation? In other words, what happens in your logic when a Talawa end user creates an event, that happens to be recurring?
1. How is it the same and different from the stackoverflow post?
    1. https://stackoverflow.com/questions/50204299/recurring-events-database-model

According to the Stack Overflow post Google calendar:

1. Creates a series of single instance events whenever an event is created
1. Makes all events in a recurring series are linked to an DB collection/table tracking the series
1. Makes modifications to an event by back referencing the tracker to update all the events in the series
    1. For example, if you add people to a calendar entry, the back reference to the tracker gets you all the calendar entries in the series. Then it updates all the individual calendar entries from day X onward with the update.
1. Stores the details of each recurrence in the tracker

This seems like a relatively simple methodology. Is that what you are planning?

We need to ensure as much backward compatibility as possible with the Talawa client apps that expect this feature to work.

**Comment by palisadoes** [comment]
I think we'll need to split this up into multiple PR with child issues:

1. Create a recurring instance defaulting to a week
2. Edit the recurring instance (invitees)
3. Create recurring instances for (separate PRs each which we could open to other contributors too)
    1. every x (weeks, months)
    2. editing recurring instances (every x weeks to every x months and vice versa, invitees)
    3. for x (weekdays, days)
    4. editing recurring instances (every x day to every x months and/or weeks and vice versa)
1. For the other custom Google options

**Comment by Community-Programmer** [comment]
> @Community-Programmer
> 
> 1. What is the logic of operation? In other words, what happens in your logic when a Talawa end user creates an event, that happens to be recurring?
> 2. How is it the same and different from the stackoverflow post?
>    
>    1. https://stackoverflow.com/questions/50204299/recurring-events-database-model
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/293441027-d309890d-1897-4fbf-8b41-7037589dc1b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM5MTExMzIsIm5iZiI6MTcwMzkxMDgzMiwicGF0aCI6Ii8xNjg3NTgwMy8yOTM0NDEwMjctZDMwOTg5MGQtMTg5Ny00ZmJmLThiNDEtNzAzNzU4OWRjMWI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjMwVDA0MzM1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxY2FmNjM5YWY4MzNiZjViMmU5Zjc2ODRjMzgzZTRjYTY2MzgxMDg5ZmY1YTcyYjY0NTFjNzBkMjJiNjkxMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8pFsqtp6QdQJc6CDW3nzKQD3CtvGtSTnbG0H_0qw5sw)
> 
> According to the Stack Overflow post Google calendar:
> 
> 1. Creates a series of single instance events whenever an event is created
> 2. Makes all events in a recurring series linked to an DB collection/table tracking the series
> 3. Makes modifications to an event by back referencing the tracker to update all the future events in the series
>    
>    1. For example, if you add people to a calendar entry, the back reference to the tracker gets you all the calendar entries in the series. Then it updates all the individual calendar entries from day X onward with the update.
> 4. Stores the details of each recurrence in the tracker
> 5. Changes in the frequency, creates a new recurring series entry starting from day X, and all the old events remain with the previous series, and the previous series ends when the new one starts.
> 
> This seems like a relatively simple methodology. Is that what you are planning?
> 
> We need to ensure as much backward compatibility as possible with the Talawa client apps that expect this feature to work.

The logic is nearly the same: when a Talawa end user creates a recurring event, the createRecurringEvent mutation is called. This mutation generates all recurring event instances from the specified start date to end date, with the user-selected frequency ["ONCE," "DAILY," "WEEKLY," "MONTHLY," "YEARLY"]. By default, the frequency is set to weekly.
If a user doesn't create a recurring event, the createEvent mutation is called.

This is the approach I am considering. I would appreciate your insights on it.

Thank You

**Comment by Community-Programmer** [comment]
> I think we'll need to split this up into multiple PR with child issues:
> 
> 1. Create a recurring instance defaulting to a week
> 2. Edit the recurring instance
>    
>    1. Change invitees
>       
>       1. All the above at any point in the recurrence with the changes propagating forward
>    2. Change end date
>    3. No end date
> 3. Create recurring instances for (separate PRs each which we could open to other contributors too)
>    
>    1. for every x (weeks, months) with and without end dates
>    2. for every x (weekdays, days) with and without end dates
>    3. editing recurring instances
>       
>       1. Every x weeks to every x months and vice versa
>       2. Invitees
>       3. End dates
>       4. All the above at any point in the recurrence with the changes propagating forward
>    4. editing recurring instances (every x days to every x months and/or weeks and vice versa)
> 4. For the other custom Google options

Ok, breaking down the tasks into multiple pull requests (PRs) with child issues is a thoughtful approach.

**Comment by palisadoes** [comment]
Why not create recurring events all the time? If it's not recurring it would just generate a single event. That way you don't need a separate mutation. 

Then if a single event were made recurring, it would be easier logic, wouldn't it?

**Comment by Community-Programmer** [comment]
I appreciate your suggestion, and it's valid. Initially, I opted for a separate mutation to enhance clarity and understanding of the codebase. However, upon further consideration, I agree that maintaining a single mutation could simplify the logic without compromising functionality.

I'll made the adjustment accordingly. Thank you for your guidance

**Comment by palisadoes** [comment]
OK. Please open a PR when ready with the appropriate tests.

I'd suggest thinking of all the various parameters to use in the mutation and implement them one at a time. 

We need to change the sample database to include weekly recurring events so that a full calendar will always be visible. This will help other developers get a sense of what we're trying to achieve. There's a separate issue for that.

Think of how we could support this first, possibly with a Talawa Admin collaborator to add a weekly option when creating an event. A checkbox to start with more UI friendly options later.

Remember we need to create incremental PRs to fix incrementally advancing functionality.

At some point we'll have multiple people working simultaneously on the mutation logic which will cause merge conflicts. Think of ways of simply refactoring the logic into  classes (or something equivalent), each residing in a separate file to reduce the risk without making the code hard to follow. Hopefully this would allow us to delegate sequencing work a class/file at a time. 

For example, classes for weeks, months, days, weekdays and all the customizations within those time frames that could occur. All tied together with the standard event generation and regeneration logic.

There may be better ways that you could think of too.

**Comment by palisadoes** [comment]
- https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github%2Fworkflows%2Finspect.yml

There are two schema generators at the end of this GitHub action file. One of them exports markdown files that we commit to the Talawa Docs repo to create the Talawa website. It will have the information you require on how to document your schema code to make the markdown user friendly

**Comment by Community-Programmer** [comment]
Sure, I'll follow your suggestions and work on implementing the mutation logic for the weekly recurring events. I'll start by considering various parameters and implementing them one at a time, as you've suggested.

I'll keep you updated on the progress and ensure that the PRs are created with appropriate tests.

Thank You

**Comment by DecodeAndCode** [comment]
Can I raise the PR after this issue is resolved.

**Comment by palisadoes** [comment]
@Community-Programmer Please resubmit the PR as it was merged without realizing that there were reviewer comments

**Comment by Community-Programmer** [comment]
@palisadoes  Ok, I appreciate the feedback. I will resubmit the PR with the suggested changes as per the reviewer's comments.

**Comment by palisadoes** [comment]
Ask the same reviewers to do the reviews for the new one

**Comment by palisadoes** [comment]
This issue needs to be tackled by someone with the skills and abilities mentioned in the issue description. 

1. This is not for a newbie
1. This is not for someone with limited time over the next month

**Comment by meetulr** [comment]
I'd like to work on this.

**Comment by meetulr** [comment]
@palisadoes I'll be starting on this issue today. I request some time to go through all the discussions and the code.
I understand it's an important issue. I'll make it a priority.
I'll reach out if I have any doubts or suggestions.

Thank you!

**Comment by palisadoes** [comment]
How's progress on the analysis?

**Comment by meetulr** [comment]
> How's progress on the analysis?

Yes, I went through everything, I was just typing out the approach as you made this comment. I'll post it soon.

**Comment by meetulr** [comment]
Alright... after going through the discussions and blogs, this is what I'd suggest:

## Issue:
### *What do we have?*
1. We have the functionalities to deal with single non-recurring events.

### *What do we need?*
1. We need the ability to create recurring events.
2. We need the functionality to implement custom recurring patterns, like top calendar apps out there.
3. We need the ability to update and delete:
- a single instance of the recurring pattern
- this and the future instances
- all instances.
5. We need a way to track the historical record for an event.

## Solution:
### *Schemas/Interfaces*
![RecurringEvents](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/24269e41-f8e0-433c-b3dc-78e82e46069a)

The purpose and need for each of the fields and schemas will be explained in the approach as their necessity arises.

### *Approach*
1. We will use the rrule libary and follow the dynamic generation approach.

#### *Creating recurring events*
1. First, let's talk about the input:
- The EventInput would have `recurring` field set to true, but about the `recurrence` field that we currently have, we should remove it and add an `rrule` field instead which would be a string.
- The recurrence field in the existing approach is an enum : `ONCE`, `WEEKLY`, `MONTHLY`, etc... But the rrule already has that in it's [`freq`](https://github.com/jkbrzt/rrule#rrule-constructor), and we want an exact pattern for event creation.
-  We want to have google calendar like functionality. So we should be implementing modals similar to that in the frontend to select recurring patterns.
- We should use rrule in the frontend too, which, according to the values we select, and the boxes we check, would easily create an rrule string for us in the frontend and we could send it to the backend in the input.

2. After getting the input, we'd follow these steps:
- Create a RecurrenceRule document that would contain the rrule sent from the frontend, let's name this document's _id to be `recurranceRuleId`.
- Create a TrueRecurringEvent that would just be like creating a normal event, let's name it's _id to be `trueRecurringEventId` (This is what we will use as the base event for generating instances.)
- Both of these would contain the `startDate` and the `endDate` as provided in the input.
- We would decide on an endDate, say `X` months or years ahead from the current date, that would decide the number of instances we create in the createEvent mutation, or during the query.
- Get the dates for the instances to be generated using the `rrule`. Here:
  - If the event has no `endDate`, we'd generate events upto `X` date.
  - If the `endDate` is very far into the future, we'd generate events upto `X` date, and leave the rest for the query.
  - If there is an `endDate`, and the last date returned from the rrule is less than `X`, then we just create all the instances.
- All the instances (Event documents) we created in the previous step will be based on the `TrueRecurringEvent` document we created above. All of them would have their *recurrenceRuleId* field set to `recurranceRuleId`, and the *trueRecurringEventId* set to `trueRecurringEventId` (from the first two steps).
- Update the `latestInstanceDate` of the `RecurrenceRule` document created to be equal to the start date of the last instance we generated here.


#### *Updating recurring events*
1. ##### Updating this instance only:
- This would be straightforward, just make a regular update.

2. ##### Updating all instances / this and future instances:

    Here things would be a little complicated, and i would need your opinion on this. I think we should base it on the rrule:
- If the rrule has not changed:
  - then we can just make a regular update.
  - update the `TrueRecurringEvent` according the update inputs (which would be used as the new base event).
- If the rrule has changed:
  - Now correct me if i'm wrong here, but the user would change the rrule for current and future events only, not the past events that has already occured (changing their recurrence pattern wouldn't make sense, any changes other than rrule that are made on them would be covered by the first point).
  - If so, we would follow these steps:
    - Find the last instance that was generated before the current instance, say `latestInstance`, and set the `latestInstanceDate` and the `endDate` of the `recurringRule` for the current pattern to be the `latestInstance's` `startDate`.
    - Delete every instance (current and the future) conforming to the current rrule. We can do this because we are generating events dynamically, i.e. we are only creating instances upto a certain date, so not many documents have to be deleted.
    - Update the `TrueRecurringEvent` document to have values of the current update input (which would be used as the new base event).
    - Create a new `RecurrenceRule` document with the new `rrule` from the input. Set it's `startDate` as the current date, and `endDate` as the `TrueRecurringEvent's` `endDate`. Now, follow the same steps that were used in creating events in createEvent mutation to generate instances.
    - Now, all the previous instances would have a different `RecurrenceRule` than the current and future ones. 

  What I'm suggesting here is that when the user changes the rrule and hits "save", this and the future instances will be affected.
  *Note here that we're not creating a new TrueRecurringEvent document, just updating the existing one.*

#### *Deleting recurring events*
1. ##### Deleting this instance only:
- Make a regular deletion.

2. ##### Deleting all instances / this and future instances:
- For deleting all instances:
  - Delete all the recurring instances with the current `trueRecurringEventId`. (So you see, `TrueRecurringEvent`, aside from being used as the base event to create new instances, also connects all the instances, even if their `rrule` are different. Which means we could also use it to track the historical record of a recurring event, accross all the instances, no matter what recurrence pattern it followed at any point).

- For this and future instances:
  - Find the document that was created previously to the current document with the current `RecurrenceRule`, set the `latestInstanceDate` and the `endDate` of the `RecurrenceRule` to that instance's `startDate`. Update the `TrueRecurringEvent` accordingly (modifying the `endDate`).
  - Delete all the recurring instances with the same `recurringRuleId` as the current document, starting from the current date (The way google calendar does it).

  
#### *Querying events*

Currently we're just querying all the events belonging to an organization. We would need significant modifications in the query or make a seperate query for this.
These are the steps we'd follow:
  - Find all the events upto the date `X`(what we'd decided on earlier) ahead of the current date.
  - Find all the `RecurrenceRule` documents with the `endDate` greater than `&&` `latestInstanceDate` less than `X`.
  - Find the `TrueRecurringEvent`s for those.
  - Generate the recurring instances upto `X`.
  - Update the `latestInstanceDate` of the `RecurrenceRule` documents.
  - Return all the documents we fetched/generated here.

#### *Historical References*
As mentioned earlier, `TrueRecurringEvent` will take care of that.


### One other approach here could be:
- Getting rid of the `TrueRecurringEvent`, instead using an `isException: boolean` field and a `trueEventId: (uuid string)` field in the event schema. Using this approach, we won't have a base event for generating new instances, so this is how we'd handle it:
  - When making an update on a single instance, we could mark it as an exception (`isException: true`).
  - When generating new instances, we'd first have to find the last instance with the given `RecurrenceRule` that isn't an exception, and then create instances based on that. But what if there's a case where all the instances are made an exception? Is that possible?
  - The `trueEventId` with the unique `uuid` here could be used for connecting recurring instances accross different `rrule`s and tracking historical records.

I would prefer using the `TrueRecurringEvent`.

@palisadoes @EshaanAgg @xoldyckk Please take a look and tell me If I've missed anything, or if any modifications are required.
Thank you!!

**Comment by palisadoes** [comment]
1. Make all events recurring. A single event would just have 0 repetitions. That streamlines the algorithm. We discussed that in the first PR for this issue that was subsequently merged. For example the NewEvent could have:
    1. the new recurring parameters plus an embedded OldEvent (what we have now) as the template to be used when generating events dynamically. 
    1. a way to store single exceptions as a list of OldEvents (see below)
1. When modifying a recurrence you would:
    1. create two almost duplicate recurring events:
         1. one with the only difference being that the start date would be the end date of the modified original recurrence that you are about to delete. This helps to maintain the original continuity beyond the end date.
         1. another with the only difference being that the end date would be the start date of the modified original recurrence that you are about to delete. This helps to maintain the original continuity before the  date.
    1. create a new recurring event for the desired end and start date. (You can verify this behavior with Google Calendar. I gave some examples in the PR that was closed.)
    1. However RRULE has a way of handling exceptions when only one instance in a recurrence is edited. There was some discussion of that too. This is where the list of OldEvents would be stored. I think you could use the RRULE to generate the dates off the events and then check the list of OldEvent exceeptions to overrule the embedded OldEvent in NewEvent. (Coming to think of it, this exception approach may have been discussed in one of the many links in either of the PRs. I'm not sure)
1. We also didn't want the RRULE stored in the field as a string, but as the fields in the RRULE class constructor, possibly with more intuitive names. It would be a little more difficult but it could allow us to replace the RRULE  package more easily in future if required (my opinion, though unlikely). It would also make the interpretation of a DB document easier during development and troubleshooting as the RRULE strings are cryptic (this will be helpful when we have little documentation and many developers over the years). @EshaanAgg had other reasons.
1. Deleting recurring events would be handled in a similar way to modifications, with recurring events start and end dates being modified before and after the period of deletion with adjustments for the lists of exceptions.
1. You mentioned that you'd prefer not to have exceptions, so in that case you'd just create more atomized recurrences. There is a disadvantage to this approach because you would need to manage a linked list of recurring events so that you could continue to edit the recurrence from the day of the exception onward. It adds a little more complexity. Once again you should test what Google does in such a case.

**Comment by palisadoes** [comment]
I forgot about the idea of creating N future Event instances with queries. It was decided upon before, so that would been to be accommodated. 

There would be possible Event instance deletions with some edits if there were past Event queries over a period overlapping the RRULE modification that would have created Events.

**Comment by meetulr** [comment]
![1st point](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/1077829c-8766-4f32-b858-86014ce5db43)
1. Sure, we could default the recurrence pattern to have a count of 1, on that day of the week.
2. I suggested the idea of `TrueRecurringEvent` because it takes care of the exceptions, and us having to store a list for them seperately. We can generate the new instances based on this `TrueRecurringEvent`, which is updated with every update in all/future instances, so we would be generating the new instances with the latest values. I don't understand the need for storing exceptions as a list. Because they could just be edited seperately, and the new generated instances won't be affected by it.
3. We're worried about the instances that have been modified on their own (by updating "this instance only") right? They're not affecting the new instance generation in any way. And when some earlier instance is modified with ("this and the future events or all events"), they'd follow that too, every instance belonging to an rrule would be affected (like google calendar).

![2nd point](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/dba8ec22-da07-41a2-914a-37e482397068)
1. That's what the method I suggested does, creating the new `RecurrenceRule`, modifying the start and end dates for the instances following different recurrence patterns, and modifying the base event, to generate new instances based on the latest information.

![3rd point](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/5be14b64-8ed8-4640-a289-a417fcb742ac)
1. I think this wasn't made very clear in the discussion. Here we would have `RecurrenceRule` that would be a single document for many instances, so we can store the rrule fields seperately, would make interpretation easy as you said.
2. About the possibility of not relying on rrule in the future, that would be a whole lot of work on its own, using a string or having fields seperately won't be a concern there, main work would be the implementing the logic.

![5th point](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/7c2bb769-f681-4fe3-a205-d16ab2345449)
1. My approach is motivated from google calendar itself, when we edit the recurrence pattern there, it creates a new series, starting from that date, and when we make updates ("all / this and future") on previous instances, the new series isn't affected, that's what my method does too.
2. The whole idea is that exceptions could be there for instances conforming to an rrule, but when bulk edits or deletes are done, the rrule wins. That's what I'm suggesting, and that's what google calendar does. There is no linked list here, bulk updations are being done based on the changes in rrule.

**Comment by meetulr** [comment]
My understanding of the exceptions was such that, among the scheduled instances, we could move instances some days forward or backwards, changes it's attendees list, etc.., that would make it an exception, and the new instances to be generated should not be affected by that (that's what `TrueRecurringEvent` is for).
I overlooked that we could have exceptions that would, say, on deletion of the whole recurring pattern, would not be deleted (i.e. excluded them from the rule). In that case, we should add the `isException` field in the `Event` schema. Through which we could do operations on whole patterns while excluding the documents where `isException: true`. If we want them to conform again to the rrule, we could update the `isException: false`.

**Comment by meetulr** [comment]
> I forgot about the idea of creating N future Event instances with queries. It was decided upon before, so that would been to be accommodated.
> 
> There would be possible Event instance deletions with some edits if there were past Event queries over a period overlapping the RRULE modification that would have created Events.

I explained the dynamic generation of events in the query process in my solution.

I think my suggestion would make it simple for future developers to understood and worked on. Could you explain exactly where it fails to align with the functionalities we want.

**Comment by palisadoes** [comment]
@meetulr 

Please make sure you can address this issue when creating new events. I think the last implementation of recurring events added the requirement of a replication set, which isn't the default behavior if you are running MongoDB on `localhost`.

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/65b33f1e-0423-452a-843f-f367f1e8eb0d)

**Comment by palisadoes** [comment]
I see. `TrueRecurringEvent` is the proposed way to handle recurring events, which would generate a limited set of instances of future Events depending on the RRULE parameters. 

`TrueRecurringEvent` is for explanatory purposes here, but it should be renamed to `RecurringEvent` to prevent future confusion.

```
RecurrenceRule
|-> TrueRecurringEvent
  |-> Event x N
```
My interpretation of your logic is that:

1. The first N events are created at once all linked back to `TrueRecurringEvent`. 
2. If a single event N - x is later edited in any way, it wouldn't change its linkage back to `TrueRecurringEvent`, so all events already created could be easily modified.
3. If event # 1 is changed, for all future events, then updating all the pre-existing events would be easy because of the linkage.
4. If the recurrence is extended, you'd apply the rule to events past the N date. This means you'll have to store the N date, in case the N date instance's date or time is modified as an exception. You could create two N date entries each with different times on the same day, or that may be a risk we wish to take.
5. When you make an edit of date `Y and beyond`, then you have to create a new `RecurrenceRule` and possibly do some deletions.
6. You'll need to think of a way to handle scrolling through calendars after instance N and auto-generating more instances and updating the N date.
7. You'll need some error detection logic if the clients are sending RRULE strings. Some sort of checksum and possibly using the npm library's `rrulestr` function to correctly create an RRULE object without error as an extra step. That would make it easier for you to store the RRULE parameters versus the string as you would have known for sure it passed the test.

Is this correct?

**Comment by meetulr** [comment]
> 1. The first N events are created at once all linked back to TrueRecurringEvent.

> 2. If a single event N - x is later edited in any way, it wouldn't change its linkage back > to TrueRecurringEvent, so all events already created could be easily modified.

> 3. If event # 1 is changed, for all future events, then updating all the pre-existing 
> events would be easy because of the linkage.
Yes, all events conforming to an rrule can be updated (except for the ones with `isException: true`) because of the `RecurrenceRule` linkage. The `TrueRecurrenceEvent` is the base event for generating new instances and its linkage is for tracking historical records.

> 4. If the recurrence is extended, you'd apply the rule to events past the N date. This      > means you'll have to store the N date, in case the N date instance's date or time is 
> modified as an exception. You could create two N date entries each with different times > on the same day, or that may be a risk we wish to take.
The date upto which the instances have been generated would be represented by the `latestInstanceDate` (which would be equal to the `startDate` of the last instance that was generated during the query or during the first `createEvent` mutation) of the `RecurrenceRule`.
So, when new instance generation is required, we could check it's `latestInstanceDate` and `endDate` against the date upto which we want to create instances in this query (as i explained in the `Quering Events` part in my original suggestion, as suggested [here](https://github.com/PalisadoesFoundation/talawa-api/pull/1712#issuecomment-1903644746) by @EshaanAgg ). So, once again, no need for exceptions here.
The `startDate` and `endDate` of the `RecurringEvent` and `TrueRecurringEvent` would initially be equal to the input we provided. But if `rrule`s are changed, those properties would change in the `RecurrenceRule` document, not in the `TrueRecurringEvent`. Because once modified, different `rrules` would be treated as seperate entities (like google calendar).
The `endDate` of the `TrueRecurringEvent` would only be modified when we delete all the future events conforming to the latest `rrule`.

> 5. When you make an edit of date Y and beyond, then you have to create a new 
> RecurrenceRule and possibly do some deletions.
Yes.

> 6. You'll need to think of a way to handle scrolling through calendars after instance N > and auto-generating more instances and updating the N date.
Yes, for that I was thinking we could have the current calendar date as a parameter in the query, so that depending on that date, we could fix a future date in the query upto which we want to create instances during this request.

> 7. You'll need some error detection logic if the clients are sending RRULE strings. 
> Some sort of checksum and possibly using the npm library's rrulestr function to 
> correctly create an RRULE object without error as an extra step. That would make it 
> easier for you to store the RRULE parameters versus the string as you would have 
> known for sure it passed the test.
That's why I was suggesting we use `rrule` in the frontend too, based on the values we select through the modals, it would easily create a valid `rrule` string, developers wouldn't have to worry about that in the frontend.

**Comment by meetulr** [comment]
I also added this part in my original suggestion:
![updateTrueRecurringEvent](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/4eee25e5-1f5b-4d1c-a5a7-8ff70d857812)

**Comment by meetulr** [comment]
> @meetulr
> 
> Please make sure you can address this issue when creating new events. I think the last implementation of recurring events added the requirement of a replication set, which isn't the default behavior if you are running MongoDB on `localhost`.
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/302064847-65b33f1e-0423-452a-843f-f367f1e8eb0d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwMTg3MzEsIm5iZiI6MTcwNzAxODQzMSwicGF0aCI6Ii8xNjg3NTgwMy8zMDIwNjQ4NDctNjViMzNmMWUtMDQyMy00NTJhLTg0M2YtZjM2N2YxZThlYjBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDAzNDcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2ZjI2NmZmMDc5YThmZjQ4Mzk4Y2Y0OWVkMDIyNDBlNzBjMDM5ZWVkMmUwMDQ5NTIxM2YxZTlmODRkZjdjMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JoEoW5ImoTdlgIiDVAKeT8Acftea6N2P5aC6rFx07Vc)

I'm confused about this. We need transactions, right? If so, we'd need a replica set for it to work. The issue #1772  is about that right?

**Comment by palisadoes** [comment]
> > @meetulr
> > Please make sure you can address this issue when creating new events. I think the last implementation of recurring events added the requirement of a replication set, which isn't the default behavior if you are running MongoDB on `localhost`.
> > ![image](https://private-user-images.githubusercontent.com/16875803/302064847-65b33f1e-0423-452a-843f-f367f1e8eb0d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwMTg3MzEsIm5iZiI6MTcwNzAxODQzMSwicGF0aCI6Ii8xNjg3NTgwMy8zMDIwNjQ4NDctNjViMzNmMWUtMDQyMy00NTJhLTg0M2YtZjM2N2YxZThlYjBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDAzNDcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2ZjI2NmZmMDc5YThmZjQ4Mzk4Y2Y0OWVkMDIyNDBlNzBjMDM5ZWVkMmUwMDQ5NTIxM2YxZTlmODRkZjdjMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JoEoW5ImoTdlgIiDVAKeT8Acftea6N2P5aC6rFx07Vc)
> 
> I'm confused about this. We need transactions, right? If so, we'd need a replica set for it to work. The issue #1772 is about that right?

Yes #1772

**Comment by palisadoes** [comment]
@meetulr 

1. I understand how the exception flag works, without the need for an exception list.
2. You are implying that the only time the RRULE string will be used is over the network by an encode/decode mechanism, everywhere else the RRULE object would be used.
1. I think it would be best to split this into several smaller PRs because the concepts  though not hard, are closely interrelated. It will be easy to lose track. 
    2. Think of ways to do this affecting no more than about 20 files at a time.
    3. Start on the first one when you have a plan. No need for confirmation as long as the 20 file limit is met.

The solution should work. Kudos to @Community-Programmer and @EshaanAgg for getting us most of the way.

**Comment by meetulr** [comment]
Alright... I'll make a plan for PRs. I think I'd be better if @EshaanAgg and @xoldyckk  vet this approach once too.

**Comment by meetulr** [comment]
> The solution should work. Kudos to @Community-Programmer and @EshaanAgg for getting us most of the way.

You too sir😅.

**Comment by palisadoes** [comment]
There was also @xoldyckk who kept us true to finding the best long term solution even when I had my doubts that we could stray. Which is what we hopefully have done now. Kudos too.

**Comment by meetulr** [comment]
@palisadoes I can just add these 4 fields in the event schema:
![eventInterface](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/f8ded863-b3a2-41e6-b69a-0b96dec9126b)
Instead of creating a seperate collection for `TrueRecurringEvent` (`RecurringEvent` as you suggested should be named).
Would this be favourable? If so, I think the name is correct, otherwise it would cause confusion with "recurring" field. Or I could name it `BaseRecurringEvent`?
The approach would be the same.
Please let me know.

**Comment by palisadoes** [comment]
1. `IsException` should be `isRecurringEventException`
1. Why do you need `recurringEventId`when there is already `_id` for the Event? 
1. Why do you need `isRecurringEvent` when all events will be recurring? Single events will have no repetitions.

**Comment by meetulr** [comment]
> 2. Why do you need recurringEventIdwhen there is already _id for the Event?

Are you talking about the `recurrenceRuleId`? It's for the `RecurrenceRule` document linkage.


> 3. Why do you need isRecurringEvent when all events will be recurring? Single events will have no repetitions.

I meant if we're not creating a seperate collection for `TrueRecurringEvent` model identical to `Event` model, we would need a way to seperate the `TrueRecurringEvent` from a normal event, so that it won't be returned during normal queries, can be queried for its own operations with ease, with just a `isTrueRecurringEvent` flag.

**Comment by palisadoes** [comment]
1. Thanks for the clarification. Yes, we'll need a `BaseRecurringEvent` I didn't realize that this was separate from Event in your description.
2. If you make all events recurring, then the methodology would create a `BaseRecurringEvent` for  a single event, but that would make it easier to make the single event into a multi-instance event later. It would also create a single algorithm for handling all events. This was the goal when I said all events should be recurring.

**Comment by meetulr** [comment]
Actually, on second thought, I think we should keep them seperate, it'd be more intuitive that way, I'll be making a `generateRecurringInstances` function that would be utilized in the `createEvent`, `updateEvent` mutations, and also the `query` (baically wherever we might need to generate instances).

> 2. If you make all events recurring, then the methodology would create a BaseRecurringEvent for a single event, but that would make it easier to make the single event into a multi-instance event later. It would also create a single algorithm for handling all events. This was the goal when I said all events should be recurring.

We shouldn't create a `BaseRecurringEvent` by default for every normal event. Many single non-recurring events would not be made recurring, that means we'd have extra documents (`BaseRecurringEvent`) that we've no use of. This is what I'm doing:
![eventSchema](https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/991b7273-518d-4a03-9b0e-7e6bb5db77d9)

We'd only generate `BaseRecurringEvent` and `RecurrenceRule` documents for recurring events.

If a non-recurring event is made recurring later on, we could easily create a `BaseRecurringEvent` and a `RecurrenceRule` document for that, that would be handled in the `updateEvent` mutation, and from there we'd call the `generateRecurringInstances` function.

The code for normal events is straightforward and intuitive as it is, I think mixing it with events for which `recurring: true` and same functions (`generateRecurringInstances`) would create some confusion.
Having 'recurring` field makes more sense to me. Also, it would be more intuitive on the frontend, we could toggle options to show modals (to select recurrence patterns, like google calendar) based on this field.

What do you think?

**Comment by palisadoes** [comment]
OK, please proceed with this approach

**Comment by palisadoes** [comment]
Option 2. Wasn't this what we discussed previously?

**Comment by meetulr** [comment]
Right, I was thinking about it in terms of current/future or past event.

Typical user behavior while updating a single event to be recurring would be for two cases:
1. Some past single event is made recurring, in which case the user would provide a `startDate` (current or future date) and possibly, an `endDate`, and the instances would be created on the dates generated by the recurrence rule.
2. Some current or future event is made recurring, like mistakenly creating a recurring event as a single event, and then updating it to be a recurring event, or just converting a single event (current/future) into a recurring event. In such cases, if the `startDate` is not provided, or is the same as the event's `startDate`, then the recurrence rule could generate a date that coincides with the existing event, and we'd have two same events on that day. That's why I was asking to remove the existing one. But I asked in haste, I forgot about the past events being made recurring.

No worries though, I could add an argument that would advance the date to one day ahead while we're updating the single event to be recurring, which would cover both cases.

Please let me know if I've misunderstood.

**Comment by meetulr** [comment]
Right, I was thinking about it in terms of current/future or past event.

Typical user behavior while updating a single event to be recurring would be for two cases:
1. Some past single event is made recurring (like an experimental event agreed to be made recurring), in which case the user would provide a `startDate` (current or future date) and possibly, an `endDate`, and the instances would be created on the dates generated by the recurrence rule.
2. Some current or future event is made recurring, like mistakenly creating a recurring event as a single event, and then updating it to be a recurring event, or just converting a single event (current/future) into a recurring event. In such cases, if the `startDate` is not provided, or is the same as the event's `startDate`, then the recurrence rule could generate a date that overlaps with the existing event, and we'd have two same events on that day. Or the current event's date might not conform to the recurrence pattern used to convert the single event to be recurring. That's why I was asking if we want to delete the existing event, and just generate a new series, which would seem appropriate in this case, I think.

Please let me know if I've misunderstood.

**Comment by palisadoes** [comment]
1. To reduce duplication, deletion of a single event converted to recurring is one approach.
2. Why couldn't you just ignore the expected duplication with the regeneration?

**Comment by meetulr** [comment]
> 1. To reduce duplication, deletion of a single event converted to recurring is one approach.
> 2. Why couldn't you just ignore the expected duplication with the regeneration?

I'm currently calling the `createRecurringEvent` directly from the `updateSingleEvent` function, which handles everything. It would call the `generateDates` function which generates dates based on startDate and endDate provided in the argument.

To avoid the duplication, i could add an extra `generateAhead` input argument for `generateDates` which would be true when updating single event to be recurring, and based on that we could shift the recurrence start date one day ahead, recurrenceRule would not be changed, so that the dates would be generated starting from the next day, avoiding duplication.

But this wouldn't take care of the future event that is made recurring but it's startDate doesn't conform to the recurrenceRule provided.
For that we can add a `keepEvent` field in the `UpdateEventInput`, which together with `generateAhead`, would take care of all these scenarios.

I was asking because these inputs aren't that intuitive😅. But for handling these edge cases they would be required. Would have to document them properly.

**Comment by palisadoes** [comment]
1. We need an intuitive approach even at the expense of minor inefficiency. If that can be achieved with the deletion method, then do that.
2. Many people will be editing this code later. Explain the rationale in the comments if necessary

**Comment by palisadoes** [comment]
Do it the Google way

**Comment by palisadoes** [comment]
What else is outstanding?

**Comment by meetulr** [comment]
> What else is outstanding?

Delete functionality. After that, basic CRUD for recurring events would be done.

**Comment by meetulr** [comment]
Currently, when we delete single events with `removeEvent` mutation, we simply update their status to 'DELETED', and when we  delete it with `adminRemoveEvent` mutation, we're deleting the documents from the database too.

Should we do the same when deleting multiple instances of a recurring event?

**Comment by palisadoes** [comment]
1. Why the inconsistency between deletions by Admins and regular users? It should be the same. Is it an oversight? If there is no undelete feature, then they should all be deleted.
2. I don't think there is a `BLOCKED` status that is used, therefore that only leaves `ACTIVE` which therefore could be completely removed

**Comment by meetulr** [comment]
> Why the inconsistency between deletions by Admins and regular users? It should be the same. Is it an oversight? If there is no undelete feature, then they should all be deleted.

That's what I wanted to know.

I asked because separate mutations were made and tested. What was the original purpose of the `status` field??
Were there initially some plans that justified implementing it this way?
The event interface describes status as `ACTIVE`, `BLOCKED`, or `DELETED`.

> I don't think there is a `BLOCKED` status that is used, therefore that only leaves `ACTIVE` which therefore could be completely removed.

Yes, it's not being used, and neither is `adminRemoveEvent`, `removeEvent` is being used, which changes the status of the event to `DELETED`.

The query for events currently returns all the "ACTIVE" events.

**Comment by Community-Programmer** [comment]
@meetulr 

I hope this message finds you well. I wanted to express my deepest gratitude for stepping in and handling this issue while I am deeply engrossed in my exams. Your willingness to take on this responsibility means more to me than words can convey
Thank you so much!!

**Comment by palisadoes** [comment]
1. We should remove the orphan code, especially if `DELETED` and `BLOCKED` is never used by the clients.
1. Verify with the other repos and create issues to make the updates.

Keeping the DELETED flag adds to the confusion as in certain cases the recurring events are physically deleted, versus just having the flag being set.

**Comment by meetulr** [comment]
I think the `status` is not being used anywhere.

So, I'm deleting the documents (from database) when deleting an event, rather than changing its `status`.

We could fix it later if it causes some breaking change.

**Comment by palisadoes** [comment]
OK

**Comment by meetulr** [comment]
> @meetulr
> 
> I hope this message finds you well. I wanted to express my deepest gratitude for stepping in and handling this issue while I am deeply engrossed in my exams. Your willingness to take on this responsibility means more to me than words can convey Thank you so much!!

Ofcourse brother, your research and discussion on this issue played a major part. It helped a lot!

**Comment by meetulr** [comment]
> OK. Remove the status code too.

I have removed it from the delete event mutation (For single events too).

**Comment by palisadoes** [comment]
1. Should be closed now?
2. What about the documentation for Talawa-Docs?

**Comment by meetulr** [comment]
> 1. Should be closed now?

Yes

> 2. What about the documentation for Talawa-Docs?

Yes, I'll open an issue in `talawa-docs`.


Can I open an issue for recurring events in admin? I could start implementing features, and make corrections to the backend (if required) along the way.

**Comment by meetulr** [comment]
Can I open an issue for recurring events in admin? I could start implementing the features, and make corrections to the backend (if required) along the way.

**Comment by palisadoes** [comment]
Yes
